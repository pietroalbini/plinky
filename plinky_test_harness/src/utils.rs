use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
