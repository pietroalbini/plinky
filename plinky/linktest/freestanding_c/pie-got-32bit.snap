---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  section .text#0 (perms: rx) in hello.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 89 e5 53 83 ec 04 e8 fc ff ff ff 81 c3 02 00 │ U..S............ │
 │   │  │ 00 00 83 ec 04 6a 0c 8d 83 00 00 00 00 50 6a 01 │ .....j.......Pj. │
 │   │  │ ff 93 00 00 00 00 83 c4 10 83 ec 0c 6a 00 ff 93 │ ............j... │
 │   │  │ 00 00 00 00                                     │ ....             │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭───────────────────────┬───────────────────────┬────────┬────────╮
 │   │  │ Type                  │ Symbol                │ Offset │ Addend │
 │   │  ├───────────────────────┼───────────────────────┼────────┼────────┤
 │   │  │ Relative32            │ __x86.get_pc_thunk.bx │ 0x8    │        │
 │   │  ├───────────────────────┼───────────────────────┼────────┼────────┤
 │   │  │ GOTLocationRelative32 │ _GLOBAL_OFFSET_TABLE_ │ 0xe    │        │
 │   │  ├───────────────────────┼───────────────────────┼────────┼────────┤
 │   │  │ OffsetFromGOT32       │ <section .rodata>     │ 0x19   │        │
 │   │  ├───────────────────────┼───────────────────────┼────────┼────────┤
 │   │  │ GOTIndex32            │ write                 │ 0x22   │        │
 │   │  ├───────────────────────┼───────────────────────┼────────┼────────┤
 │   │  │ GOTIndex32            │ exit                  │ 0x30   │        │
 │   │  ╰───────────────────────┴───────────────────────┴────────┴────────╯
 │   ┴
 │
 │  section .text#1 (perms: rx) in syscalls.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 53 51 b8 04 00 00 00 8b 5c 24 10 8b 4c 24 14 │ PSQ......\$..L$. │
 │   │  │ 8b 54 24 18 cd 80 59 5b 58 c3 b8 01 00 00 00 8b │ .T$...Y[X....... │
 │   │  │ 5c 24 04 cd 80                                  │ \$...            │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 ┴

debug print: calculated layout
 │
 │    Layout:
 │  ╭─────────────────────────────┬─────────────┬─────────────┬────────────────┬───────────────╮
 │  │ Part                        │ File offset │ File length │ Memory address │ Memory length │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <elf header>                │ 0x0         │ 0x34        │ 0x0            │ 0x34          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <program header>            │ 0x34        │ 0x100       │ 0x34           │ 0x100         │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .dynstr                     │ 0x134       │ 0xd         │ 0x134          │ 0xd           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .dynsym                     │ 0x141       │ 0x30        │ 0x141          │ 0x30          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .hash                       │ 0x171       │ 0x18        │ 0x171          │ 0x18          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .dynamic                    │ 0x189       │ 0x50        │ 0x189          │ 0x50          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .rela.got                   │ 0x1d9       │ 0x18        │ 0x1d9          │ 0x18          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .interp                     │ 0x1f1       │ 0x13        │ 0x1f1          │ 0x13          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <section header>            │ 0x204       │ 0x2d0       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .shstrtab                   │ 0x4d4       │ 0x9d        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .strtab                     │ 0x571       │ 0x51        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .symtab                     │ 0x5c2       │ 0x80        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .comment#2                  │ 0x642       │ 0x3b        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>                   │ 0x67d       │ 0x983       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .rodata                     │ 0x1000      │ 0xd         │ 0x1000         │ 0xd           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .eh_frame                   │ 0x100d      │ 0x48        │ 0x100d         │ 0x48          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>                   │ 0x1055      │ 0xfab       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text#0                     │ 0x2000      │ 0x34        │ 0x2000         │ 0x34          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text.__x86.get_pc_thunk.bx │ 0x2034      │ 0x4         │ 0x2034         │ 0x4           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text#1                     │ 0x2038      │ 0x25        │ 0x2038         │ 0x25          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>                   │ 0x205d      │ 0xfa3       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .got                        │ 0x3000      │ 0x8         │ 0x3000         │ 0x8           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .got.plt                    │ 0x3008      │ 0xc         │ 0x3008         │ 0xc           │
 │  ╰─────────────────────────────┴─────────────┴─────────────┴────────────────┴───────────────╯
 │
 │    Segments:
 │  ╭────────┬────────┬────────────────┬───────┬─────────────────────────────╮
 │  │ Start  │ Align  │ Type           │ Perms │ Content                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x34   │ 0x1000 │ program header │ R     │ <program header>            │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x1f1  │ 0x1    │ interpreter    │ R     │ .interp                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x0    │ 0x1000 │ program        │ R     │ <elf header>                │
 │  │        │        │                │       │ <program header>            │
 │  │        │        │                │       │ .dynstr                     │
 │  │        │        │                │       │ .dynsym                     │
 │  │        │        │                │       │ .hash                       │
 │  │        │        │                │       │ .dynamic                    │
 │  │        │        │                │       │ .rela.got                   │
 │  │        │        │                │       │ .interp                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x1000 │ 0x1000 │ program        │ R     │ .rodata                     │
 │  │        │        │                │       │ .eh_frame                   │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x2000 │ 0x1000 │ program        │ RX    │ .text#0                     │
 │  │        │        │                │       │ .text.__x86.get_pc_thunk.bx │
 │  │        │        │                │       │ .text#1                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x3000 │ 0x1000 │ program        │ RW    │ .got                        │
 │  │        │        │                │       │ .got.plt                    │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x189  │ 0x4    │ dynamic        │ R     │ .dynamic                    │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ -      │ 0x1    │ GNU stack      │ RW    │                             │
 │  ╰────────┴────────┴────────────────┴───────┴─────────────────────────────╯
 │
 │    deduplication facade .comment#0 in hello.o
 │  ╭──────┬──────────────────╮
 │  │ From │ To               │
 │  ├──────┼──────────────────┤
 │  │ 0x0  │ .comment#2 + 0x0 │
 │  ├──────┼──────────────────┤
 │  │ 0x1  │ .comment#2 + 0x1 │
 │  ╰──────┴──────────────────╯
 │
 │    deduplication facade .comment#1 in <plinky>
 │  ╭──────┬───────────────────╮
 │  │ From │ To                │
 │  ├──────┼───────────────────┤
 │  │ 0x0  │ .comment#2 + 0x2c │
 │  ╰──────┴───────────────────╯
 ┴

debug print: object after relocations are applied
 │
 │  section .got (perms: rw) in <plinky>
 │   │
 │   │  address: 0x3000
 │   │
 │   │  ╭─────────────────────────┬──────────╮
 │   │  │ 00 00 00 00 00 00 00 00 │ ........ │
 │   │  ╰─────────────────────────┴──────────╯
 │   ┴
 │
 │  section .got.plt (perms: rw) in <plinky>
 │   │
 │   │  address: 0x3008
 │   │
 │   │  ╭─────────────────────────────────────┬──────────────╮
 │   │  │ 89 01 00 00 00 00 00 00 00 00 00 00 │ ............ │
 │   │  ╰─────────────────────────────────────┴──────────────╯
 │   ┴
 │
 │  section .interp (perms: r) in <plinky>
 │   │
 │   │  address: 0x1f1
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 2f 6c 69 62 2f 6c 64 2d 6c 69 6e 75 78 2e 73 6f │ /lib/ld-linux.so │
 │   │  │ 2e 32 00                                        │ .2.              │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#0 (perms: rx) in hello.o
 │   │
 │   │  address: 0x2000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 89 e5 53 83 ec 04 e8 28 00 00 00 81 c3 f4 0f │ U..S....(....... │
 │   │  │ 00 00 83 ec 04 6a 0c 8d 83 00 e0 ff ff 50 6a 01 │ .....j.......Pj. │
 │   │  │ ff 93 00 00 00 00 83 c4 10 83 ec 0c 6a 00 ff 93 │ ............j... │
 │   │  │ 04 00 00 00                                     │ ....             │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#1 (perms: rx) in syscalls.o
 │   │
 │   │  address: 0x2038
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 53 51 b8 04 00 00 00 8b 5c 24 10 8b 4c 24 14 │ PSQ......\$..L$. │
 │   │  │ 8b 54 24 18 cd 80 59 5b 58 c3 b8 01 00 00 00 8b │ .T$...Y[X....... │
 │   │  │ 5c 24 04 cd 80                                  │ \$...            │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭───────────────────────────────────────┬──────────┬────────────┬────────────┬───────────────────────────────────╮
 │  │ Name                                  │ Type     │ Source     │ Visibility │ Value                             │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ <empty>                               │ none     │ <plinky>   │ local      │ <null>                            │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ <section .text#0>                     │ section  │ hello.o    │ local      │ .text#0 + 0x0                     │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ <section .rodata>                     │ section  │ hello.o    │ local      │ .rodata + 0x0                     │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ <section .text.__x86.get_pc_thunk.bx> │ section  │ hello.o    │ local      │ .text.__x86.get_pc_thunk.bx + 0x0 │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ _DYNAMIC                              │ none     │ <plinky>   │ local      │ .dynamic + 0x0                    │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ _GLOBAL_OFFSET_TABLE_                 │ none     │ <plinky>   │ local      │ .got.plt + 0x0                    │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ __x86.get_pc_thunk.bx                 │ function │ hello.o    │ local      │ .text.__x86.get_pc_thunk.bx + 0x0 │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ _start                                │ function │ hello.o    │ global     │ .text#0 + 0x0                     │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ exit                                  │ none     │ syscalls.o │ global     │ .text#1 + 0x1a                    │
 │  ├───────────────────────────────────────┼──────────┼────────────┼────────────┼───────────────────────────────────┤
 │  │ write                                 │ none     │ syscalls.o │ global     │ .text#1 + 0x0                     │
 │  ╰───────────────────────────────────────┴──────────┴────────────┴────────────┴───────────────────────────────────╯
 │
 │    Dynamic symbols:
 │  ╭─────────┬──────┬────────────┬────────────┬────────────────╮
 │  │ Name    │ Type │ Source     │ Visibility │ Value          │
 │  ├─────────┼──────┼────────────┼────────────┼────────────────┤
 │  │ <empty> │ none │ <plinky>   │ local      │ <null>         │
 │  ├─────────┼──────┼────────────┼────────────┼────────────────┤
 │  │ exit    │ none │ syscalls.o │ global     │ .text#1 + 0x1a │
 │  ├─────────┼──────┼────────────┼────────────┼────────────────┤
 │  │ write   │ none │ syscalls.o │ global     │ .text#1 + 0x0  │
 │  ╰─────────┴──────┴────────────┴────────────┴────────────────╯
 ┴

debug print: built elf
 │
 │    Metadata:
 │  ╭────────────┬───────────────╮
 │  │ Class      │ ELF 32bit     │
 │  ├────────────┼───────────────┤
 │  │ Endian     │ Little        │
 │  ├────────────┼───────────────┤
 │  │ ABI        │ System V      │
 │  ├────────────┼───────────────┤
 │  │ Machine    │ x86           │
 │  ├────────────┼───────────────┤
 │  │ Type       │ Shared object │
 │  ├────────────┼───────────────┤
 │  │ Entrypoint │ 0x2000        │
 │  ╰────────────┴───────────────╯
 │
 │  section .dynstr#10 (address: 0x134)
 │   │
 │   │    Strings table:
 │   │  ╭─────┬───────╮
 │   │  │ 0x0 │       │
 │   │  ├─────┼───────┤
 │   │  │ 0x1 │       │
 │   │  ├─────┼───────┤
 │   │  │ 0x2 │ write │
 │   │  ├─────┼───────┤
 │   │  │ 0x8 │ exit  │
 │   │  ╰─────┴───────╯
 │   ┴
 │
 │  section .dynsym#11 (address: 0x141)
 │   │
 │   │    Dynamic symbol table:
 │   │  ╭─────────┬─────────┬──────┬────────────┬────────────┬────────┬──────╮
 │   │  │ Name    │ Binding │ Type │ Visibility │ Definition │ Value  │ Size │
 │   │  ├─────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ #8      │ Local   │ -    │ Default    │ Undefined  │ 0x0    │ 0x0  │
 │   │  ├─────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ write#9 │ Global  │ -    │ Default    │ .text#6    │ 0x2038 │ 0x0  │
 │   │  ├─────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ exit#10 │ Global  │ -    │ Default    │ .text#6    │ 0x2052 │ 0x0  │
 │   │  ╰─────────┴─────────┴──────┴────────────┴────────────┴────────┴──────╯
 │   ┴
 │
 │  section .hash#12 (address: 0x171)
 │   │
 │   │  Hash table for .dynsym#11
 │   │
 │   │    Content:
 │   │  ╭───────────┬───────────────────╮
 │   │  │ Bucket ID │ Symbols in bucket │
 │   │  ├───────────┼───────────────────┤
 │   │  │ 0         │ exit#10           │
 │   │  │           │ write#9           │
 │   │  ╰───────────┴───────────────────╯
 │   ┴
 │
 │  section .dynamic#13 (address: 0x189)
 │   │
 │   │  dynamic information | string table: .dynstr#10
 │   │
 │   │  ╭─────────────────────────┬───────────────╮
 │   │  │ Kind                    │ Value         │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ String table            │ address 0x134 │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ String table size       │ 13 bytes      │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ Symbol table            │ address 0x141 │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ Symbol table entry size │ 16 bytes      │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ Hash table              │ address 0x171 │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ RelocationsA table      │ address 0x1d9 │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ RelocationsA size       │ 24 bytes      │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ RelocationsA entry size │ 12 bytes      │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ Flags1                  │ pie           │
 │   │  ├─────────────────────────┼───────────────┤
 │   │  │ Null                    │ -             │
 │   │  ╰─────────────────────────┴───────────────╯
 │   ┴
 │
 │  section .rela.got#15 (address: 0x1d9)
 │   │
 │   │  symbol table:       .dynsym#11
 │   │  applies to section: .got#14
 │   │
 │   │    Relocations:
 │   │  ╭─────────────┬─────────┬────────┬────────╮
 │   │  │ Type        │ Symbol  │ Offset │ Addend │
 │   │  ├─────────────┼─────────┼────────┼────────┤
 │   │  │ X86_GlobDat │ write#9 │ 0x3000 │ 0x0    │
 │   │  ├─────────────┼─────────┼────────┼────────┤
 │   │  │ X86_GlobDat │ exit#10 │ 0x3004 │ 0x0    │
 │   │  ╰─────────────┴─────────┴────────┴────────╯
 │   ┴
 │
 │  section .interp#17 (address: 0x1f1)
 │   │
 │   │  program data | permissions: R
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 2f 6c 69 62 2f 6c 64 2d 6c 69 6e 75 78 2e 73 6f │ /lib/ld-linux.so │
 │   │  │ 2e 32 00                                        │ .2.              │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │    Segments:
 │  ╭──────────────────────┬───────┬──────────┬─────────────┬──────────┬────────────────┬────────────╮
 │  │ Type                 │ Perms │ Aligment │ File offset │ File len │ Memory address │ Memory len │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Program header table │ R     │ 0x1000   │ 0x34        │ 0x100    │ 0x34           │ 0x100      │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Interpreter          │ R     │ 0x1      │ 0x1f1       │ 0x13     │ 0x1f1          │ 0x13       │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                 │ R     │ 0x1000   │ 0x0         │ 0x204    │ 0x0            │ 0x204      │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                 │ R     │ 0x1000   │ 0x1000      │ 0x55     │ 0x1000         │ 0x55       │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                 │ R X   │ 0x1000   │ 0x2000      │ 0x5d     │ 0x2000         │ 0x5d       │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                 │ RW    │ 0x1000   │ 0x3000      │ 0x14     │ 0x3000         │ 0x14       │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Dynamic              │ R     │ 0x4      │ 0x189       │ 0x50     │ 0x189          │ 0x50       │
 │  ├──────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ GNU stack            │ RW    │ 0x1      │ 0x0         │ 0x0      │ 0x0            │ 0x0        │
 │  ╰──────────────────────┴───────┴──────────┴─────────────┴──────────┴────────────────┴────────────╯
 ┴

==============

running exited with exit status: 0

=== stdout ===
Hello world

no stderr present
